Implement the build repository script in the buildchain

ycointe opened the pull request:
**Component**: Build

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: MetalK8s on Ubuntu

**Summary**: We want to implement the script which will create Debian repositories

**Acceptance criteria**:
Launch the ISO build with `./doit.sh` then verify that we have all the Debian repositories in `_build/root/packages/debian/`


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #1746 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

ycointe committed:
build/packaging: Change the Debian packages location

When building repositories, the packages are copied into the repo
hierarchy.

Until now, we were downloading package on the ISO directly (under
`_build/root`) but since the repo copy them, we shouldn't download them
here (they will be duplicated).

This commit updates the code to download the packages in the "staging"
area (under `_build/packages`).

Refs: #1746

bert-e commented:
# Hello ycointe,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @slaperche-scality

* @gdemonet

* @carlito-scality

* @sayf-eddine-scality

* @nootal

* @thomasdanan

* @alexandre-allard-scality

* @lucieleonard

* @zempashi

* @monpote

* @jbertran

* @cmonfort

* @nicolast

* @teddyandrieux

slaperche-scality commented on buildchain/buildchain/packaging.py:
```diff
@@ -176,28 +176,28 @@ def clean() -> None:
     """Download Debian packages locally."""
     def clean() -> None:
-        """Delete cache and repositories on the ISO."""
+        """Delete local Debian packages."""
```
`Delete downloaded Debian packages`

slaperche-scality commented on buildchain/buildchain/packaging.py:
```diff
@@ -176,28 +176,28 @@ def clean() -> None:
             if repository.packages:
                 continue
-            coreutils.rm_rf(repository.rootdir)
+            coreutils.rm_rf(constants.PKG_DEB_ROOT/repository.fullname)
```
I would add:

```python
def pkgdir(self) -> Path:
    return constants.PKG_DEB_ROOT/self.fullname
```

in the class `DEBRepository` and the use `repository.pkgdir` here and l.192

slaperche-scality commented on buildchain/buildchain/packaging.py:
```diff
@@ -211,7 +211,7 @@ def mkdirs() -> None:
         'title': utils.title_with_target1('GET DEB PKGS'),
         'actions': [mkdirs, dl_packages_callable],
-        'targets': [constants.REPO_DEB_ROOT/'.witness'],
+        'targets': [constants.PKG_DEB_ROOT/'.witness'],
```
Commit message should explain why better, like:

```
When building repositories, the packages are copied into the repo hierarchy.
Until now, we were downloading package on the ISO directly (under `_build/root`) but since the repo copy them, we shouldn't download them here (they will be duplicated).
This commit updates the code to download the packages in the "staging" area (under `_build/packages`).
```

Something like that.

ycointe committed:
build/repository: Move _mkdir_repo_root in Repository

We move this function because it is used by the both OS families, so we
can factorize it to put it in the parent class `Repository`.

Refs: #1746

ycointe committed:
build/repository: Build repo from downloaded packages

We defined a task to call the entrypoint function `buildrepo` which
build Debian repository from Debian packages.

Refs: #1746

ycointe committed:
build/package: Change the location of build packages

We change the location of built packages. We put the in `_build/packages`
to use them later to make easier the metalk8s-scality repository build.

Refs: #1746

ycointe committed:
build/package: Move code down to RPMPackage

As we don't use the function `make_package_directory` for Debian, we
removed it from `Package` to only put it in `RPMPackage`.

Refs: #1746

ycointe committed:
build/repository: Build repo from built packages

Permit to build repository for manually built packages like
metalk8s-sosreport and calico-cni-plugin.

Refs: #1746

ycointe committed:
build/packaging: Add deb repo build in packaging target

Add a task which permits to call all the functions required to build all
the Debian repositories.

Refs: #1746

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @sayf-eddine-scality

* @thomasdanan

* @carlito-scality

* @teddyandrieux

* @nootal

* @lucieleonard

* @gdemonet

* @jbertran

* @nicolast

* @monpote

* @tcarmet

* @slaperche-scality

ycointe committed:
build/utils: Move the function to remove files

We move this function in `utils` to have to pissibility to use it
somewhere else.

ycointe committed:
build/package: Use the function to remove files

We need this function to delete files to permit to do a
`./doit.sh clean`

jbertran commented on buildchain/buildchain/targets/package.py:
```diff
@@ -349,7 +349,7 @@ def __init__(
         """DEB path."""
         fmt = '{pkg.name}_{pkg.version}-{pkg.build_id}_{pkg.ARCH}.deb'
-        return self.rootdir/fmt.format(pkg=self)
+        return constants.PKG_DEB_ROOT/fmt.format(pkg=self)
```
In the final staging area, this looks odd since we have every other `.deb` that was downloaded set inside its directory (so `_build/packages/metalk8s-bionic/*.deb` etc), but the built packages are set in `_build/packages`.

Could we make that `constants.PKG_DEB_ROOT / 'metalk8s-scality' / fmt.format(pkg=self)` ?

ycointe replied:
It will be handled by the issue #1781

jbertran commented on buildchain/buildchain/targets/package.py:
```diff
@@ -375,7 +372,7 @@ def build_package(self) -> types.TaskDict:
             ),
             utils.bind_mount(
-                source=self.rootdir, target=Path('/debbuild/result')
+                source=constants.PKG_DEB_ROOT, target=Path('/debbuild/result')
```
From my comment above, this one would change as well.

ycointe replied:
It will be handled by the issue #1781

jbertran commented on buildchain/buildchain/targets/package.py:
```diff
@@ -404,14 +401,10 @@ def build_package(self) -> types.TaskDict:
-            ),
+            utils.bind_ro_mount(source=self.sources,
+                                target=Path('/rpmbuild/source.rpm')),
+            utils.bind_mount(source=constants.PKG_DEB_ROOT,
```
And this one.

ycointe replied:
It will be handled by the issue #1781

jbertran commented on buildchain/buildchain/targets/repository.py:
```diff
@@ -331,10 +331,14 @@ def fullname(self) -> str:
     def pkgdir(self) -> Path:
         """Repository where to download the packages."""
+        if self.packages:
+            # Built packages are not under a sub-directory.
```
But they should be, I guess.

ycointe replied:
It will be handled by the issue #1781

jbertran reviewed (commented):
A comment on the built packages staging area.

From what I gather you handled the `.witness` deletion issue while I was reviewing.

ycointe commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye ycointe.
